messaging/src/main/java/org/axonframework/lifecycle/ShutdownLatch.java

```diff
@@ -0,0 +1,98 @@
+package org.axonframework.lifecycle;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+/**
+ * A latch implementation to be used in shutdown scenarios. Operations to wait for can be added by invoking {@link
+ * #increment()}. An added operation should always be removed through {@link #decrement()} once it has completed, as
+ * {@link #await()} will otherwise block indefinitely. If the latch is waited on through {@link #await()}, new
+ * operations can no longer be added.
+ *
+ * @author Steven van Beelen
+ * @since 4.3
+ */
+public class ShutdownLatch {
+
+    private final AtomicInteger operationCounter = new AtomicInteger(0);
+    private CompletableFuture<Void> latch;
+
+    /**
+     * Add an operation this latch should wait on before opening up. If this operation is called whilst {@link #await()}
+     * is already called an {@link IllegalStateException} will be thrown.
+     *
+     * @throws IllegalStateException if {@link #await()} has been called prior to invoking this method
+     */
+    public void increment() {
```

Comment:
To make this method completely safe in its usage, it's best if this method returns a "handle" that can be closed to indicate that a certain workload has been finished. The method signature could look like:
`public ActivityHandle registerActivity()`. ActivityHandler would have a single method: `void end()`.
This implementation of `ActivtyHandle.end()` could also verify that it wasn't invoked more than once. Decrementing the counter more than incrementing it can be problematic.

Reply:
Brilliant idea to make this even safer. Gonna do something like this.